### PR TITLE
Expand snocket tests

### PR DESCRIPTION
**Pull request by bolt12**

### CODE REVIEW

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (coot) [inline ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs:797]: Let's rename it as `runClient`.

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (coot) [inline ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs:761]: Let's rename it as `runServer`.

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (coot) [inline ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs:767]: Let's add a haddock.

We dont need to pass `Snocket` to it, it is the same as the one passed in the previous argument.

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (coot) [inline ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs:804]: As in the other function:
* add a haddock
* we don't need to pass `Snocket` to the handler.

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (coot) [inline ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs:768]: Add a haddock as well.

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (coot) [inline ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs:815]: This function deserves an explanation (haddocks).

I'd suggest another name: `listenAndConnect`

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (coot) [inline ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs:820]:
```suggestion
 => TestAddress addr -- ^ local address
 -> TestAddress addr -- ^ remote address
```

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (coot) [inline ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs:787]: Let's rename it as `acceptOne`.

The boolean value is redundant, it would be better to return `Either fd IOError`, but `Maybe fd` is also fine if we don't need the error (e.g. for a counter example).

**Review by coot — COMMENTED** [review]
Some comments, I'll continue reviewing next week.

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (bolt12) [reply ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs:208]: Yes I missed this bit, will change!

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (bolt12) [reply ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs:224]: Might have been some name clashing during development, I'll change this! Thanks!

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (coot) [inline ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs:714]:
```suggestion
        case fromException err
               :: Maybe (ResourceException (TestAddress Int)) of
```

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (coot) [inline ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs:728]: `loop` seems to be only used in this function, I would move it the `where` close here.

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (coot) [inline ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs:749]: Just for completeness sake, could we also add a test which connects to a socket that is bound but is not a listening?

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (coot) [inline ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs:753]: The same comment as in `prop_connect_to_accepting_socket` applies here as well.

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (coot) [inline ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs:683]: I think we should verify that the connection was established by looking at the net-sim state, rather than just check that there's no error.

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (bolt12) [reply ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs:667]: I am not sure I understood what you meant.

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (bolt12) [reply ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs:786]: Good question, the tests didn't trigger any. But I'll look closer

Comment (bolt12) [reply ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs:786]: since this is inside a `bracket` if an async exception triggers then the cleanup function from the top-level `bracket` will run, right?

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (bolt12) [reply ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs:683]: I'll rebase this PR on #3417 to get access to `UniverseState`

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (coot) [reply ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs:376]: If we have `counterexample (f1 tr) (counterexample (f2 tr) ...)` we must ensure that nighter `f1` nor `f2` throws an exception during evaluation.  The `selectTraceEventsSay'` (and friends) are used for that purpose; but if we avoid multiple `counterexamples` in the first place we didn't had this problem in the first place.  Though sometimes it might be more convenient to have multiple `counterexamples`.

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (coot) [inline ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs:477]:
```suggestion
                         snocket (close snocket)
```

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (coot) [inline ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs:494]:
```suggestion
                             snocket (close snocket)
```

Comment (coot) [inline ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs:494]:
```suggestion
                             snocket (close snocket)
```

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (coot) [inline ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs:617]: `assertUniverseState` -> `assertNetworkState`

please `grep -ri universe ouroboros-network-framework` if there are other traces of it still around.

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (coot) [inline ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs:635]:
```suggestion
      Right _ ->
```

**Review by coot — APPROVED** [review]
LGTM, I left some minor comments.

**Review by coot — CHANGES_REQUESTED** [review]
Hmm, the commit: `Added multinode_cm_counters_Sim test`, does not seem to belong to this PR.

**Comment by bolt12** [comment]
@coot that commit history means that this PR is based on top of my counters testing PR in order to get access to the `ObservableNetworkState`. Thank you for the review, I'll address it asap!

**Comment by bolt12** [comment]
I'll merge this once the counters PR has been merged.

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (coot) [reply ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs]: yes it will but still the accepted socket will not be closed.

Comment (bolt12) [reply ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs]: how so? the cleanup function runs `close` on the socket. If that runs then the socket is closed, am I confusing things?

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (coot) [inline ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs:650]: I'd move this where `NormalisedId` is defined, and not export `NormalisedId` constructor.  I would also call it `mkNormalisedId`.

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (coot) [inline ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs:636]: we should return something more descriptive than `UnexpectedOutcome`.

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (coot) [inline ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs:357]: I think it's worth to define more constructors which are used depending on the situation.

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (coot) [inline ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs:360]: If I am not mistaken we're never `throwIO` these errors, so the `Exception` instance is not really needed.  I put a comment below how to avoid `toExceptions` as well.

**Review by coot — CHANGES_REQUESTED** [review]

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (bolt12) [reply ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs:462]: Ahh much nicer!

Comment (bolt12) [reply ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs]: fixed!

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (bolt12) [reply ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs]: fixed!

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (bolt12) [reply ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs]: fixed!

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (bolt12) [reply ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs]: fixed!

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (bolt12) [reply ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs]: fixed!

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (bolt12) [reply ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs]: fixed!

File: ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs
Comment (coot) [reply ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs]: I see, you're right.

**Review by coot — APPROVED** [review]
LGTM

**Comment by bolt12** [comment]
bors merge

**Comment by bolt12** [comment]
bors merge

**Pull request merged**
